src/hdmf/common/resources.py

```diff
@@ -0,0 +1,357 @@
+                     'the object_id for the Container/Data object that uses the key')},
+            {'name': 'field', 'type': str, 'doc': 'the field of the Container/Data that uses the key', 'default': None},
+            {'name': 'key', 'type': (str, Key), 'default': None,
+             'doc': 'the name of the key or the Row object from the KeyTable for the key to add a resource for'},
+            {'name': 'resource_name', 'type': str, 'doc': 'the online resource (i.e. database) name', 'default': None},
+            {'name': 'entity_id', 'type': str, 'doc': 'the identifier for the entity at the resource', 'default': None},
+            {'name': 'entity_uri', 'type': str, 'doc': 'the URI for the identifier at the resource', 'default': None})
+    def add_ref(self, **kwargs):
+        """
+        Add information about an external reference used in this file.
+
+        It is possible to use the same *key_name* to refer to different resources so long as the *key_name* is not
+        used within the same object and field. This method does not support such functionality by default. The different
+        keys must be added separately using *add_key* and passed to the *key* argument in separate calls of this method.
+        """
+        container = kwargs['container']
+        field = kwargs['field']
+        key = kwargs['key']
+        resource_name = kwargs['resource_name']
+        resource_id = kwargs['entity_id']
+        resource_uri = kwargs['entity_uri']
+        add_rsc = False
+
+        if resource_name is not None and resource_id is not None and resource_uri is not None:
+            add_rsc = True
+        elif not (resource_name is None and resource_id is None and resource_uri is None):
+            msg = ("Specify all or none of resource_name, entity_id, and entity_uri arguments. "
+                   "All three are required to create a reference")
+            raise ValueError(msg)
+
+        if isinstance(container, Container):
+            container = container.object_id
+
+        object_field = None
+
+        # get Key object by searching the table
+        if not isinstance(key, Key):
+            key_id = self.keys.which(key_name=key)
+
+            if len(key_id) == 0:
+                # the key has never been used before
+                key = self.add_key(key)
+
+            elif len(key_id) > 1:
+                # if same key is used multiple times, determine
+                # which instance based on the Container
+                object_field = self._check_object_field(container, field)
+
+                key_tmp = self.object_keys['keytable_idx', object_field.idx]
+                if key_tmp in key_id:
+                    key = self.keys.row[key_tmp]
+                else:
+                    key = self.add_key(key)
+            else:
+                key = self.keys.row[key_id[0]]
+
+        if object_field is None:
+            object_field = self._check_object_field(container, field)
```

Comment:
Logic comment -- since object_field is set to this no matter what, can line 266 be changed to 
`object_field = self._check_object_field(container, field)` ?